### PR TITLE
[cga][cudamapper] STL-like allocator and buffer, helper arrays also allocated using custom allocators

### LINKED ISSUE #292: [cga] Adapt custom allocator and buffer to be more in line with STL
DeviceAllocator should have `T* DeviceAllocator<T>::allocate()` instead of `void* DeviceAllocator::allocate()`. This is required in order to use DeviceAllocator for internal temporary arrays in Thrust's algorithms.
Doing so will enable us to allocate more memory for caching allocator and also reduce time Thrust spends allocating temporary arrays.

This requires substantial changes to design and adaptation of existing usages of `device_buffer` in cudamapper

### PR DESCRIPTION
**Pull request by mimaric**
Custom allocators and buffer are more STL-like (although not fully compliant).

Before this change there was only instance of custom allocator and all buffers held a shared_ptr to it. Now custom allocator holds a shared_prt to underlying cub allocator (if there is one) and every buffer holds a copy of that allocator (all copies point to the same underlying cub allocator).

Changes:
- Allocators templatized
- `allocate()` returns `pointer`, not `void*`
- `allocate()` accepts number of elements to allocate, not bytes
- `deallocate()` accepts `pointer`, not `void*`
- Largest cached bin in 2^28
- `CachingDeviceAllocator` holds a shared pointer to CUB's allocator so it can be copied
- `buffer` holds a copy of allocator, not a shared pointer to it (sharing the same CUB's allocator is handled by shared pointer in `CachingDeviceAllocator`)
- `buffer` returns it's allocator by value (this will enable implicit passing of allocators)
- `sort_by_two_keys` and `generate_anchors` reuse existing allocator instead of creating new ones
- Thrust's algorithms reuse existing allocator for internal arrays

This PR will require an update once #281 and #295 get merged in, but you can start reviewing

Closes #292

### CODE REVIEW

File: common/utils/include/claragenomics/utils/buffer.hpp [hunk 1]
```diff
@@ -185,15 +200,16 @@ class buffer
...
 
-    /// @brief This method swaps the contents of this buffer with another buffer.
-    /// @param v The buffer with which to swap.
-    void swap(buffer& v)
+    /// @brief This method swaps the contents of two buffers.
+    /// @param a One buffer.
+    /// @param b The other buffer.
+    friend void swap(buffer& a, buffer& b)
     {
-        std::swap(_data, v._data);
```
Comment (ahehn-nv) [inline common/utils/include/claragenomics/utils/buffer.hpp]: Add `noexcept` to swap.
While not critical here, I would not enforce `std::swap` in the implementation, but do
```
using std::swap;
swap(_size, v._size);
...
```
See https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Ro-custom. They actually use exactly this example. :D

File: common/utils/include/claragenomics/utils/device_buffer.hpp [hunk 2]
```diff
@@ -16,6 +16,10 @@ namespace claragenomics
...
 template <typename T>
-using device_buffer = buffer<T, DeviceAllocator>;
+#ifdef CGA_ENABLE_ALLOCATOR
+using device_buffer = buffer<T, CachingDeviceAllocator<T>>;
+#else
+using device_buffer = buffer<T, CudaMallocAllocator<T>>;
+#endif
 
-} // namespace claragenomics
+} // namespace claragenomics
```
Comment (ahehn-nv) [inline common/utils/include/claragenomics/utils/device_buffer.hpp]: any reason you removed the newline?

File: cudamapper/include/claragenomics/cudamapper/overlapper.hpp [hunk 3]
```diff
@@ -10,9 +10,10 @@
 
 #pragma once
 
-#include <thrust/device_vector.h>
-#include "index.hpp"
-#include "types.hpp"
+#include <claragenomics/cudamapper/index.hpp>
+#include <claragenomics/cudamapper/types.hpp>
```
Comment (ahehn-nv) [inline cudamapper/include/claragenomics/cudamapper/overlapper.hpp]: :+1:

File: cudamapper/src/index_gpu.cuh [hunk 4]
```diff
@@ -29,6 +28,10 @@
 #include <claragenomics/utils/mathutils.hpp>
 #include <claragenomics/utils/signed_integer_utils.hpp>
 
+////////////////
+#include <iostream>
+////////////////
```
Comment (ahehn-nv) [inline cudamapper/src/index_gpu.cuh]: Is `iostream` still needed?

File: cudamapper/src/matcher_gpu.cu [hunk 5]
```diff
@@ -13,7 +13,6 @@
 #include <cassert>
 #include <numeric>
 
-#include <thrust/scan.h>
```
Comment (ahehn-nv) [inline cudamapper/src/matcher_gpu.cu]: Isn't that needed for `thrust::transform_inclusive_scan`?

File: common/utils/include/claragenomics/utils/allocator.hpp [hunk 6]
```diff
@@ -9,66 +9,109 @@
...
+    /// @tparam U Type of rhs::value_type
+    /// @return reference to this object
+    template <typename U>
+    CudaMallocAllocator& operator=(CudaMallocAllocator<U>&& rhs)
+    {
+        return *this;
+    }
+
+    /// @brief destructor
+    virtual ~CudaMallocAllocator() {}
```
[the remark this answers is not in the document]
Comment (mimaric) [reply common/utils/include/claragenomics/utils/allocator.hpp]: No, I'll fix it

File: common/utils/include/claragenomics/utils/allocator.hpp [hunk 7]
```diff
@@ -9,66 +9,109 @@
...
-    virtual void* allocate(std::size_t n, cudaStream_t stream) = 0;
-
-    /// @brief Asynchronously deallocates host memory.
-    /// @param p      pointer to the buffer to deallocate
-    /// @param n      size of the buffer to deallocate in bytes
-    /// @param stream CUDA stream to be associated with this method.
-    virtual void deallocate(void* p, std::size_t n, cudaStream_t stream) = 0;
-};
+    /// @brief default constructor
+    CudaMallocAllocator(){};
```
[the remark this answers is not in the document]
Comment (mimaric) [reply common/utils/include/claragenomics/utils/allocator.hpp]: Yes, but I can still try using `= default` for default ones. I'll give it a try

File: cudamapper/src/index_gpu.cuh (hunk 4 above)
Comment (mimaric) [reply cudamapper/src/index_gpu.cuh]: no, `////////////////` was supposed to catch my attention :man_facepalming:

File: cudamapper/src/matcher_gpu.cu (hunk 5 above)
Comment (mimaric) [reply cudamapper/src/matcher_gpu.cu]: I believe not, that's why we are including `#include <thrust/transform_scan.h>`
Documentation example also does not include it: https://thrust.github.io/doc/group__transformed__prefixsums_ga88fdc1014d4c4a23995d1730e24ef919.html#ga88fdc1014d4c4a23995d1730e24ef919

Comment (ahehn-nv) [reply cudamapper/src/matcher_gpu.cu]: Ah right, makes sense. Did not see that there.

File: common/utils/include/claragenomics/utils/allocator.hpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (mimaric) [reply common/utils/include/claragenomics/utils/allocator.hpp]: done

File: common/utils/include/claragenomics/utils/allocator.hpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (mimaric) [reply common/utils/include/claragenomics/utils/allocator.hpp]: done

File: common/utils/include/claragenomics/utils/allocator.hpp [hunk 8]
```diff
@@ -79,33 +122,143 @@ class CudaMallocAllocator : public DeviceAllocator
 };
 
 /**
- * @brief A simple caching allocator for device memory allocations.
+ * @brief A simple caching allocator for device memory allocations
  */
-class CachingDeviceAllocator : public DeviceAllocator
+template <typename T>
+class CachingDeviceAllocator
```
[the remark this answers is not in the document]
Comment (mimaric) [reply common/utils/include/claragenomics/utils/allocator.hpp]: done

File: common/utils/include/claragenomics/utils/device_buffer.hpp (hunk 2 above)
Comment (mimaric) [reply common/utils/include/claragenomics/utils/device_buffer.hpp]: no, returned to the previous state

File: cudamapper/src/index_gpu.cuh (hunk 4 above)
Comment (mimaric) [reply cudamapper/src/index_gpu.cuh]: done

File: common/utils/include/claragenomics/utils/buffer.hpp (hunk 1 above)
Comment (mimaric) [reply common/utils/include/claragenomics/utils/buffer.hpp]: Cool, didn't know that!

done

Comment (ahehn-nv) [reply common/utils/include/claragenomics/utils/buffer.hpp]: :)

**Pull request merged**